File: 13-Graph-Algorithms/02-Topological-Sort/java/CourseSchedule.java
import java.util.*;
public class CourseSchedule {
    public static int[] findOrder(int n, int[][] prerequisites) {
        List<List<Integer>> adj = new ArrayList<>();
        int[] indegree = new int[n];
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());
        for (int[] p : prerequisites) { adj.get(p[1]).add(p[0]); indegree[p[0]]++; }
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < n; i++) if (indegree[i] == 0) q.add(i);
        int[] order = new int[n]; int idx = 0;
        while (!q.isEmpty()) {
            int u = q.poll(); order[idx++] = u;
            for (int v : adj.get(u)) if (--indegree[v] == 0) q.add(v);
        }
        return idx == n ? order : new int[]{};
    }
    public static void main(String[] args) {
        int[] order = findOrder(4, new int[][]{{1,0},{2,0},{3,1},{3,2}});
        for (int x : order) System.out.print(x + " ");
    }
}

File: 13-Graph-Algorithms/02-Topological-Sort/cpp/course_schedule.cpp
/*
 * Problem: Course Schedule II (Topological Sort)
 * Source:  LeetCode 210
 * Approach: Kahn's BFS-based topological sort using indegree.
 * Time: O(V+E), Space: O(V+E)
 * Tags: topological-sort, bfs, dag
 */
#include <bits/stdc++.h>
using namespace std;

vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
    vector<vector<int>> adj(numCourses);
    vector<int> indegree(numCourses, 0);
    for (auto& p : prerequisites) { adj[p[1]].push_back(p[0]); indegree[p[0]]++; }
    
    queue<int> q;
    for (int i = 0; i < numCourses; i++) if (indegree[i] == 0) q.push(i);
    
    vector<int> order;
    while (!q.empty()) {
        int u = q.front(); q.pop(); order.push_back(u);
        for (int v : adj[u]) if (--indegree[v] == 0) q.push(v);
    }
    return order.size() == numCourses ? order : vector<int>{};
}

int main() {
    vector<vector<int>> prereqs = {{1,0},{2,0},{3,1},{3,2}};
    auto order = findOrder(4, prereqs);
    for (int x : order) cout << x << " "; // 0 1 2 3 or 0 2 1 3
    return 0;
}
